Tests/SourceKitTests/BuildSystemTests.swift

```diff
@@ -227,6 +227,107 @@ final class BuildSystemTests: XCTestCase {
     }
   }
 
+  func testClangdDocumentFallbackWithholdsDiagnostics() {
+    guard haveClangd else { return }
+
+    let url = URL(fileURLWithPath: "/\(#function)/file.m")
+    let doc = DocumentURI(url)
+    let args = [url.path, "-DDEBUG"]
+    let text = """
+    #ifdef FOO
+    static void foo() {}
+    #endif
+
+    int main() {
+      foo();
+      return 0;
+    }
+    """
+
+    sk.allowUnexpectedNotification = false
+
+    sk.sendNoteSync(DidOpenTextDocumentNotification(textDocument: TextDocumentItem(
+      uri: doc,
+      language: .objective_c,
+      version: 12,
+      text: text
+    )), { (note: Notification<PublishDiagnosticsNotification>) in
+      // Expect diagnostics to be withheld.
+      XCTAssertEqual(note.params.diagnostics.count, 0)
+      XCTAssertEqual(text, self.workspace.documentManager.latestSnapshot(doc)!.text)
+    })
+
+    // Modify the build settings and inform the delegate.
+    // This should trigger a new publish diagnostics and we should see a diagnostic.
+    let newSettings = FileBuildSettings(compilerArguments: args)
+    buildSystem.buildSettingsByFile[doc] = newSettings
+
+    let expectation = XCTestExpectation(description: "refresh due to fallback --> primary")
+    sk.handleNextNotification { (note: Notification<PublishDiagnosticsNotification>) in
+      XCTAssertEqual(note.params.diagnostics.count, 1)
+      XCTAssertEqual(text, self.workspace.documentManager.latestSnapshot(doc)!.text)
+      expectation.fulfill()
+    }
+
+    buildSystem.delegate?.fileBuildSettingsChanged([doc: .modified(newSettings)])
+
+    let result = XCTWaiter.wait(for: [expectation], timeout: 5)
+    if result != .completed {
+      fatalError("error \(result) waiting for diagnostics notification")
+    }
+  }
+
+  func testSwiftDocumentFallbackWithholdsSemanticDiagnostics() {
+    let url = URL(fileURLWithPath: "/\(#function)/a.swift")
+    let doc = DocumentURI(url)
+    let settings = FallbackBuildSystem().settings(for: doc, .swift)!
+
+    let text = """
+    #if FOO
+    func foo() {}
+    #endif
+
+    foo()
```

Comment:
We should include a syntactic diagnostic to test the behaviour of filtering out semantic, not just that there are none.

Reply:
Done